nfd-worker: fix --label-whitelist

marquiz opened the pull request:
After commenting #306 I looked into the code and was not happy what I found (with my own work).

Unify handling of --label-whitelist in nfd-worker and nfd-master. That is,
in nfd-worker, apply the regexp filter on non-namespaced part of the
label name.

Brief history:
1. Originally the whitelist regexp was applied on the full namespaced
   label name (that would be e.g.
   'feature.node.kubernetes.io/cpu-cpuid.AVX' in the current nfd version)

2. Commit 81752b2d changed the behavior so that the regexp was applied
   on the non-namespaced part (that would be `cpu-cpuid.AVX`)

3. Commit 40918827 added support for custom label namespaces. With this
   change, the label whitelist handling diverged between nfd-worker and
   nfd-master. In nfd-master the whitelist regexp is always applied on
   the non-namespaced label name. However, in nfd-worker the whitelist
   handling is two-fold (and inconsistent): for labels in the standard
   nfd namespace regexp is applied on the non-namespaced part (e.g.
   `cpu-cpuid.AVX`, but, for labels in custom namespaces the regexp is
   applied on the full name (e.g. `example.com/my-feature`).

This patch changes nfd-worker to behave similarly to nfd-master. The
namespace part is now always omitted, which should be easier for the
users to comprehend.

Also, fixes a bug in the label name prefixing so that the name of the
feature source is not prefixed into labels with custom label namespace
(effectively mangling the intended namespace). For example, previously a
'example.com/feature' label from the 'custom' feature source would be
prefixed with the source name, mangling it to
'custom-example.com/feature'.

marquiz commented:
Forgot to add that this is one possible way to "fix" this. Alternatives include:
1. Change nfd-master to behave similarly to nfd-worker, instead
2. Revert back to the original, i.e. always apply the regexp to the fully namespaced label name

/hold

Ethyling commented:
I think we should apply the same policy for labels regardless of the namespace (default or not). 

However, I don't know if we should apply the filter on the namespace or not. I would say that applying the whitelisting on namespaces would add the possibility to filter labels as we want on each node. For example, currently we can't filter namespaces on nfd-worker, we can only do this on nfd-master, so it would allow us to limit namespaces per node. But it would add more "complexity" on this feature for the user.

What do you think?

marquiz commented:
These were the same questions I was pondering.

The reason I ended up the solution implemented here (i.e. whitelisting NOT applied on namespace) is that 
1. nfd-master already has other means of controlling the namespaces (`--extra-label-ns`). It is not exactly the same thing, but, probably enough
2. Labels in non-default namespaces are always result of user-made customization (hooks, feature files or the `custom` feature source). Thus, it's less likely that the user would want to filter out labels that she has has specifically enabled herself.

Ethyling commented:
> The reason I ended up the solution implemented here (i.e. whitelisting NOT applied on namespace) is that
> 
>     1. nfd-master already has other means of controlling the namespaces (`--extra-label-ns`). It is not exactly the same thing, but, probably enough
> 
>     2. Labels in non-default namespaces are always result of user-made customization (hooks, feature files or the `custom` feature source). Thus, it's less likely that the user would want to filter out labels that she has has specifically enabled herself.

Sounds reasonable to me. I will take a look at the code and test it.

mythi commented:
> Unify handling of --label-whitelist in nfd-worker and nfd-master. 

Maybe slightly off-topic but what's the use-case for double filtering?

marquiz commented:
> Maybe slightly off-topic but what's the use-case for double filtering?

In short: historical reasons 😄 It was "cloned" when NFD was split to master and worker.

marquiz commented:
/unhold

Ethyling commented:
LGTM but I think we should update the README to specify that the whitelist is not applied on the namespace part. WDYT?

marquiz commented:
> I think we should update the README to specify that the whitelist is not applied on the namespace part.

Agree. Added documentation on the command line help and via that also to the README. Do you think that would be sufficient?

Also rebased on top of the latest master.

Ethyling commented:
> > I think we should update the README to specify that the whitelist is not applied on the namespace part.
> 
> Agree. Added documentation on the command line help and via that also to the README. Do you think that would be sufficient?

Yes I think. Last checks and we can merge this :smile:

Ethyling commented:
/lgtm
